### ISSUE TITLE
Building the app with Deployment Target iOS 11 but Recommendation Feature working for iOS 12

### ISSUE THREAD

**Issue by dogahe** [apple/turicreate#1347]
We are working on an app that uses Recommender Systems model. I understand that the recommendation feature of the app won't work in iOS 11 and on the Simulator (because of bug https://github.com/apple/turicreate/issues/822). But we need to keep the deployment target of our app at iOS 11 but still would like our iOS 12 users to be able to take advantage of this feature. Therefore, I have added the following checks in code, before initializing the Recommender model:
```
#if !targetEnvironment(simulator)
        if #available(iOS 12.0, *) {
            let model = MyRecommender()
            // Use the model
           // ...
        }
#endif
```
So everything works fine in Simulator, both iOS 11 and 12 (i.e. it ignores the model). And on iOS 12 device which uses the model and makes recommendations. But the problem is that it crashes on iOS 11 devices even though, I have made `libRecommender.dylib` as Optional instead of Required under "Link Binary With Libraries". The error that I get is:

```
dyld: Symbol not found: _OBJC_CLASS_$_MLArrayBatchProvider
  Referenced from: /private/var/containers/Bundle/Application/.../Frameworks/libRecommender.dylib (which was built for iOS 12.0)
  Expected in: /System/Library/Frameworks/CoreML.framework/CoreML
 in /private/var/containers/Bundle/Application/.../Frameworks/libRecommender.dylib
Message from debugger: Terminated due to signal 6
```

Can you please let me know how I can use this model in iOS 12 devices, yet keep the deployment target as iOS 11 (knowing that, the recommendation feature of the app won't work on iOS 11)?

**Comment by nickjong** [comment]
Hmm, I feel like that should work. I wonder if we have to change how our dylib links CoreML to cater to this use case.

I'm assuming you get this error upon launch?

As a workaround for now, would it work to dlopen the dylib in the ios12 path, instead of making it a dependency in Xcode?

**Comment by nickjong** [comment]
Yeah, that's fair. Even though our docs state that the recommender only supports iOS 12, I think it's fair to expect the library to work properly when optionally linked into an iOS 11 app (and never invoked).

Until we can fix this, I'm speculating that you might be able to use `dlopen` to make your app only attempt to load the dylib after a runtime check that you're on iOS 12 (as opposed to optionally linking, which I believe will eagerly attempt to link when you app launch but not fail if the dylib is not there; the bug here is that our dylib IS there, but it requires the CoreML it finds to be an iOS 12 CoreML, apparently)

**Comment by dogahe** [comment]
Thanks. Trying your workaround worked. I only had to do 
```
#if !targetEnvironment(simulator)
        if #available(iOS 12.0, *) {
            if let handle = dlopen("libRecommender.dylib", RTLD_NOW) {
                let model = MyRecommender()
                // Use the model
               // ...
           }
        }
#endif
```
And remove libRecommender.dylib from Linked Frameworks and Libraries.
